### ISSUE TITLE
bug: Custom platform detection not respected

### ISSUE THREAD

**Issue by BoveFabio** [ionic-team/ionic-framework#28331]
### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#creating-an-issue).
- [X] I agree to follow the [Code of Conduct](https://ionicframework.com/code-of-conduct).
- [X] I have searched for [existing issues](https://github.com/ionic-team/ionic-framework/issues) that already report this problem, without success.

### Ionic Framework Version

v7.x

### Current Behavior

Ionic offers the possibility to customize platform detection. This is useful for example when accounting for laptops with touchscreens. (https://ionicframework.com/docs/react/platform#customizing-platform-detection-functions)

However, the Ionic app does not seem to respect the custom definitions set up this way.

### Expected Behavior

The Ionic app should respect the custom platform detection mechanisms.

### Steps to Reproduce

1. Install the dependencies and run the project (`npm start`) 
2. Check the console logs: they will show `false` for `isPlatform("mobile")` event though the platform detection hardcodes it to `true`. Conversely, on mobile devices `isPlatform("desktop")` will return `false` even though it is hardcoded to `true`.

### Code Reproduction URL

https://github.com/BoveFabio/ionic-platform-detection-bug

### Ionic Info

Ionic:

   Ionic CLI       : 7.1.1 (/Users/fabiobove/.nvm/versions/node/v18.16.0/lib/node_modules/@ionic/cli)
   Ionic Framework : @ionic/react 7.5.0

Capacitor:

   Capacitor CLI      : 5.4.2
   @capacitor/android : not installed
   @capacitor/core    : 5.4.2
   @capacitor/ios     : not installed

Utility:

   cordova-res : not installed globally
   native-run  : 1.7.3

System:

   NodeJS : v18.16.0
   OS     : macOS Unknown

**Comment by liamdebeasi** [comment]
Thanks for the issue. I am going to close this as this is not a bug in Ionic Framework. The problem here is you are setting your Ionic config in `index.tsx` and then overriding it in `App.tsx`. Calling `setupIonicReact` once resolved the issue on my end.

**Comment by BoveFabio** [comment]
Hey @liamdebeasi , thanks for the quick reply.
I agree, the reproduction repository was a bit short sighted. I then double checked our own app and made sure we don't call `setupIonicReact` twice. We do see the reported behavior anyway. I will get back to you once I find the actual issue and update the reproduction repo accordingly.

**Comment by liamdebeasi** [comment]
Thanks! Feel free to tag me when that's ready, and I can take another look.

**Comment by BoveFabio** [comment]
Hey @liamdebeasi ,
thanks for the offer!

So, while I did not manage to get the behavior in the reproduction app, I did find something. Maybe you can give me some directions on where else to look why our app does not pick up the custom platform detection.

So I've built a small component in our app that depends on the config I use in `setupIonicReact`:
```tsx
export function SetupIonicReactComponent() {
    return <IonLoading isOpen={true} message={`${isPlatform("mobile")}<br/>${isPlatform("desktop")}`} />;
}
```

When I call `setupIonicReact` with a config like
```ts
{
    innerHTMLTemplatesEnabled: false,
    platform: {
        mobile: (win) => {
            return true;
        },
        desktop: (win) => {
            return true;
        },
    },
}
```
then I get
<img width="178" alt="Screenshot 2023-10-11 at 19 28 37" src="https://github.com/ionic-team/ionic-framework/assets/11267625/524de698-92c1-424b-a506-e7b4d54ad267">

If I change the config to 
```ts
{
    innerHTMLTemplatesEnabled: true,
    platform: {
        mobile: (win) => {
            return true;
        },
        desktop: (win) => {
            return true;
        },
    },
}
```
I get
<img width="122" alt="Screenshot 2023-10-11 at 19 28 25" src="https://github.com/ionic-team/ionic-framework/assets/11267625/4bb78df1-6dee-4fc1-9861-2049a584b17e">

So my app picks up the `innerHTMLTemplatesEnabled` setting, but not the `platform` part.

So, in short: can you give me any pointers as to why apparently only parts of the config are applied when calling `setupIonicReact`?

**Comment by liamdebeasi** [comment]
I'd verify that the custom platform detecting is being set **before** the "isPlatform" call on the Loading component is being fired. If `isPlatform` is run first then Ionic will use the default platform detection settings.

**Comment by BoveFabio** [comment]
Hey @liamdebeasi ,

one final response from my side, so other people who might stumble across this problem can find it. 

I actually found one single place in our codebase where we use `isPlatform` outside the actual component, something like this:
```tsx
const isIPhone = isPlatform("ios") && !isPlatform("tablet");

export function SomeComponent() {
...
}
```

Changing it to 
```tsx
export function SomeComponent() {
  const isIPhone = isPlatform("ios") && !isPlatform("tablet");
  ...
}
```
did the trick. The very fact that I imported `SomeComponent` made the `isPlatform` function run before it was set up.

Thank you for your guidance!
